source/distributed/grid_tools.cc

```diff
@@ -0,0 +1,184 @@
+      //This is the coarsest level at which we can separate the connected components
+      typename parallel::distributed::Triangulation< dim, spacedim >::cell_iterator
+      parent_firstc = firstc;
+      //tmpc shall contain the children of parentc from which firstc descends
+      typename parallel::distributed::Triangulation< dim, spacedim >::cell_iterator
+      parent_lastc = lastc;
+
+      //First we reach the coarsest common level
+      if (parent_firstc->level() > parent_lastc->level() )
+        while ( parent_lastc->level() < parent_firstc->level() )
+          parent_lastc = parent_lastc->parent();
+      else if (parent_firstc->level() < parent_lastc->level())
+        while ( parent_lastc->level() > parent_firstc->level() )
+          parent_firstc = parent_firstc->parent();
+
+      //Now we look for the coarsest cell containing both firstc and lastc
+      //This is the coarsest level at which we can separate the connected components
+
+      while (parent_firstc != parent_lastc && parent_firstc->level()>0)
+        {
+          parent_firstc = parent_firstc->parent();
+          parent_lastc = parent_lastc->parent();
+        }
+
+
+      std::vector< BoundingBox < spacedim > > bounding_boxes;
+      if (parent_firstc == parent_lastc)
+        {
+          //If the two are the same we now create a bounding box for each child:
+          for (unsigned int c=0; c<parent_firstc->n_children(); ++c)
+            {
+              std::vector < typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator >
+              local_cells =
+                dealii::GridTools::get_active_child_cells < typename parallel::distributed::Triangulation< dim, spacedim > >
+                (parent_firstc->child(c));
+              bool no_bbox = true;
+              Point<spacedim> minp;
+              Point<spacedim> maxp;
+              for (unsigned int i=0; i< local_cells.size(); ++i)
+                {
+                  if (local_cells[i]->is_locally_owned())
+                    {
+                      if (no_bbox)
+                        {
+                          minp = local_cells[i]->center();
+                          maxp = local_cells[i]->center();
+                          no_bbox = false;
+                        }
+                      for (unsigned int v=0; v<GeometryInfo<spacedim>::vertices_per_cell; ++v)
+                        for ( unsigned int d=0; d<spacedim; ++d)
+                          {
+                            minp[d] = std::min( minp[d], local_cells[i]->vertex(v)[d]);
+                            maxp[d] = std::max( maxp[d], local_cells[i]->vertex(v)[d]);
+                          }
+                    }
+                }
+              if ( ! no_bbox)
+                bounding_boxes.push_back( BoundingBox < spacedim > (std::make_pair(minp,maxp)) );
+            }
+        }
+      else
+        {
+          for (auto cell: distributed_tria.cell_iterators_on_level(0))
```

Comment:
So the code below is exactly the same as the code above but the range of the loop is different? It would be nice if you encapsulate the code in one function instead of copy/pasting it (double the code, double the bugs).

Reply:
Indeed, you can refactor the same parts into internal function

Reply:
Ok, in the end I used another function.